File: Code/Lesson_14_code/toy_dns_server.py
"""A one-record authoritative DNS server, on UDP, no library.

This plays the part of `hera.ns.cloudflare.com` from Step 2's trace, but
for one made-up zone: bookmarks-api.local. It reads zone.json fresh on
every query, so editing that file changes the answer immediately — exactly
what an operator does in a DNS provider's dashboard. It answers A queries
for the one name in the zone and NXDOMAIN for anything else.

    python toy_dns_server.py
"""

from __future__ import annotations

import json
import socket
import struct
import sys
from pathlib import Path

ZONE_FILE = Path(__file__).parent / "zone.json"
LISTEN = ("127.0.0.1", 5300)


def encode_name(name: str) -> bytes:
    out = bytearray()
    for label in name.rstrip(".").split("."):
        out += bytes([len(label)]) + label.encode("ascii")
    out += b"\x00"
    return bytes(out)


def read_question_name(packet: bytes, offset: int = 12) -> tuple[str, int]:
    labels = []
    while True:
        length = packet[offset]
        if length == 0:
            offset += 1
            break
        offset += 1
        labels.append(packet[offset : offset + length].decode("ascii"))
        offset += length
    return ".".join(labels), offset


def build_answer(query: bytes, name: str, ip: str, ttl: int) -> bytes:
    query_id = query[:2]
    flags = struct.pack(">H", 0x8180)  # response, recursion available, no error
    _, qtype = struct.unpack(">HH", query[-4:])
    counts = struct.pack(">HHHH", 1, 1, 0, 0)  # 1 question, 1 answer
    question = query[12:]
    rdata = socket.inet_aton(ip)
    answer = (
        b"\xc0\x0c"  # pointer back to the name in the question section
        + struct.pack(">HHIH", 1, 1, ttl, len(rdata))  # TYPE A, CLASS IN
        + rdata
    )
    return query_id + flags + counts + question + answer


def build_nxdomain(query: bytes) -> bytes:
    query_id = query[:2]
    flags = struct.pack(">H", 0x8183)  # response, rcode 3 = NXDOMAIN
    counts = struct.pack(">HHHH", 1, 0, 0, 0)
    return query_id + flags + counts + query[12:]


def main() -> None:
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.bind(LISTEN)
        print(f"toy authoritative server listening on {LISTEN[0]}:{LISTEN[1]}", flush=True)
        while True:
            query, addr = sock.recvfrom(512)
            name, _ = read_question_name(query)
            zone = json.loads(ZONE_FILE.read_text())
            if name.lower() == zone["name"].lower():
                response = build_answer(query, name, zone["ip"], zone["ttl"])
                print(f"answer {name} -> {zone['ip']} ttl={zone['ttl']}", flush=True)
            else:
                response = build_nxdomain(query)
                print(f"NXDOMAIN {name}", flush=True)
            sock.sendto(response, addr)


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(0)

File: Code/js/Lesson_14_code/src/toyDnsServer.ts
// A one-record authoritative DNS server, on UDP, no library.
//
// The Python twin is toy_dns_server.py — same protocol, same zone.json,
// read fresh on every query. Recognition, not new material.
//
//   node toyDnsServer.ts

import dgram from "node:dgram";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ZONE_FILE = path.join(path.dirname(fileURLToPath(import.meta.url)), "zone.json");
const LISTEN_PORT = 5301;

function readQuestionName(packet: Buffer): [string, number] {
  const labels: string[] = [];
  let offset = 12;
  for (;;) {
    const length = packet[offset];
    if (length === 0) {
      offset += 1;
      break;
    }
    offset += 1;
    labels.push(packet.subarray(offset, offset + length).toString("ascii"));
    offset += length;
  }
  return [labels.join("."), offset];
}

function buildAnswer(query: Buffer, ip: string, ttl: number): Buffer {
  const id = query.subarray(0, 2);
  const flags = Buffer.from([0x81, 0x80]);
  const counts = Buffer.from([0, 1, 0, 1, 0, 0, 0, 0]);
  const question = query.subarray(12);
  const rdata = Buffer.from(ip.split(".").map(Number));
  const answerHeader = Buffer.alloc(10);
  answerHeader.writeUInt16BE(1, 0); // TYPE A
  answerHeader.writeUInt16BE(1, 2); // CLASS IN
  answerHeader.writeUInt32BE(ttl, 4);
  answerHeader.writeUInt16BE(rdata.length, 8);
  return Buffer.concat([
    id,
    flags,
    counts,
    question,
    Buffer.from([0xc0, 0x0c]), // pointer back to the question's name
    answerHeader,
    rdata,
  ]);
}

function buildNxdomain(query: Buffer): Buffer {
  const id = query.subarray(0, 2);
  const flags = Buffer.from([0x81, 0x83]); // rcode 3
  const counts = Buffer.from([0, 1, 0, 0, 0, 0, 0, 0]);
  return Buffer.concat([id, flags, counts, query.subarray(12)]);
}

const socket = dgram.createSocket("udp4");
socket.on("message", (query, remote) => {
  const [name] = readQuestionName(query);
  const zone = JSON.parse(fs.readFileSync(ZONE_FILE, "utf8"));
  if (name.toLowerCase() === zone.name.toLowerCase()) {
    socket.send(buildAnswer(query, zone.ip, zone.ttl), remote.port, remote.address);
    console.log(`answer ${name} -> ${zone.ip} ttl=${zone.ttl}`);
  } else {
    socket.send(buildNxdomain(query), remote.port, remote.address);
    console.log(`NXDOMAIN ${name}`);
  }
});
socket.bind(LISTEN_PORT, "127.0.0.1", () => {
  console.log(`toy authoritative server listening on 127.0.0.1:${LISTEN_PORT}`);
});
